stackslib/src/clarity_vm/database/mod.rs

```diff
@@ -570,6 +579,16 @@ impl BurnStateDB for SortitionHandleTx<'_> {
 }
 
 impl BurnStateDB for SortitionDBConn<'_> {
+    fn get_tip_burn_block_height(&self) -> Option<u32> {
+        let tip = SortitionDB::get_canonical_burn_chain_tip(self.conn()).ok()?;
+        tip.block_height.try_into().ok()
+    }
+
+    fn get_tip_sortition_id(&self) -> Option<SortitionId> {
+        let tip = SortitionDB::get_canonical_burn_chain_tip(self.conn()).ok()?;
+        Some(tip.sortition_id)
+    }
```

Comment:
I'm not 100% sure that this is safe. I think the `SortitionDBConn` implementation is used by the mining path, while the `SortitionHandleTx` is used by the block processor. I think that this leads to the "correct" behavior: the `SortitionHandleTx` has its chain_tip field set to the "burn_view", and when mining, the canonical tip _should_ be the tip (but I'm not sure that it always will be -- maybe a sortition could be processed between the block starting mining and finishing mining?). Either way, this seems like something that could come back to bite us someday (I think we've been bitten before by a difference between the `SortitionDBConn` and `SortitionHandleTx` impls).

My suggestion is to replace `impl BurnStateDB for SortitionDBConn` with `impl BurnStateDB for SortitionHandleConn` (which would require updating any consumers of this impl to create handles). I think this is safer in the long run (and `HandleConn` and `HandleTx` are close enough to each other, that maybe at some point, we use either a macro or the `SortitionHandle` trait to specify the exact same `impl` for both).

Reply:
Ok, I pushed a change for this, but it's possible that I may have gone overboard and let this extend out further than necessary. I'll re-review this change tomorrow.

Reply:
Also, `SortitionDB::get_canonical_burn_chain_tip()` is _NOT_ safe to use in consensus code.  Both `SortitionDBConn` and `SortitionHandleTx` have `self.context.sortition_id`, which represents their respective views of the burnchain tip.  They're guaranteed not to change over the course of their respective lifetimes, whereas the result of `SortitionDB::get_canonical_burn_chain_tip()` can change from call to call from outside of the coordinator thread.

Reply:
Furthermore, even within the coordinator thread, `SortitionDB::get_canonical_burn_chain_tip()` can return arbitrarily different values between Nakamoto blocks, which would lead to a chain split If used in this manner.  For example, suppose tenure _T_ had 10 blocks, labeled _T0_ through _T9_.  It's entirely possible that node A processes a new burnchain block in-between processing _T3_ and _T4_, while node B processes the same new burnchain block between processing _T5_ and _T6_.  Now, nodes A and B may diverge due to different return values from this impl.

Reply:
To further elaborate, the _only_ safe place to call `SortitionDB::get_canonical_burn_chain_tip()` in consensus code is within code reachable from either `ChainsCoordinator::handle_new_epoch2_burnchain_block()` or `ChainsCoordinator::handle_new_nakamoto_burnchain_block()`.  Only in these code paths is the value returned by `SortitionDB::get_ccanonical_burn_chain_tip()` guaranteed to be consistent, because these are the only two places where new sortitons can get created.

Reply:
The only new places where `get_canonical_burn_chain_tip` is called now are in tests.